### PR TITLE
Pin Python dependency versions

### PR DESCRIPTION
**Pull request by venkat2469**

### CODE REVIEW

File: tensorflow/tools/ci_build/release/requirements_common.txt [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
-protobuf ~= 3.20.3  # NOTE: Earliest version for Python 3.10
+protobuf == 4.22.3  # NOTE: Earliest version for Python 3.10
 six ~= 1.16.0
-termcolor ~= 2.1.1
-typing_extensions ~= 3.10.0.0
-wheel ~= 0.38.1
+termcolor == 2.3.0
+typing_extensions == 4.5.0
+wheel == 0.40.0
 wrapt ~= 1.14.1
```
Comment (geetachavan1) [inline tensorflow/tools/ci_build/release/requirements_common.txt:19]: Need to be ==

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt [hunk 2]
```diff
@@ -4,22 +4,22 @@
...
-absl-py ~= 1.0.0
+absl-py == 1.4.0
 astunparse ~= 1.6.3
-flatbuffers ~= 2.0
+flatbuffers == 23.3.3
 google_pasta ~= 0.2
 h5py ~= 3.8.0 # Earliest version for Python 3.11
 # TODO(b/262592253): Support older versions of NumPy for Python 3.10 and lower
 # to support TFX. Remove when Apache Beam upgrades to newer NumPy.
 numpy ~= 1.22.0; python_version < '3.11'
```
Comment (geetachavan1) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:14]: Need to be ==

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt [hunk 3]
```diff
@@ -4,22 +4,22 @@
...
-protobuf ~= 3.20.3
+protobuf == 4.22.3
 six ~= 1.16.0
-termcolor ~= 2.1.1
-typing_extensions ~= 3.10.0.0
-wheel ~= 0.38.1
+termcolor == 2.3.0
+typing_extensions == 4.5.0
+wheel == 0.40.0
 wrapt ~= 1.14.1
```
Comment (geetachavan1) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:23]: Need to be ==

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt [hunk 4]
```diff
@@ -4,22 +4,22 @@
...
 # TODO(b/262592253): Support older versions of NumPy for Python 3.10 and lower
 # to support TFX. Remove when Apache Beam upgrades to newer NumPy.
 numpy ~= 1.22.0; python_version < '3.11'
-numpy ~= 1.23.2; python_version >= '3.11' # Earliest version for Python 3.11
+numpy == 1.24.3; python_version >= '3.11' # Earliest version for Python 3.11
 opt_einsum ~= 3.3.0
 packaging ~= 21.3
-protobuf ~= 3.20.3
+protobuf == 4.22.3
 six ~= 1.16.0
```
Comment (geetachavan1) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:19]: Need to be ==

File: tensorflow/tools/ci_build/release/requirements_common.txt [hunk 5]
```diff
@@ -1,21 +1,21 @@
 # To have reproducible builds, these dependencies should be pinned always.
 # Prefer pinning to the same version as in setup.py for now.
 # This will change in the future.
-absl-py ~= 1.0.0
+absl-py == 1.4.0
```
Comment (mihaimaruseac) [inline tensorflow/tools/ci_build/release/requirements_common.txt:4]: Note that this is outside of range of previous spec. `~= 1.0.0` means `>= 1.0.0, < 1.1` but `1.4.0` is out of this range.

We might be lucky and things would work, we might be unlucky and get breakages if we use this high version.

File: tensorflow/tools/ci_build/release/requirements_common.txt [hunk 6]
```diff
@@ -1,21 +1,21 @@
...
 astunparse ~= 1.6.3
```
Comment (mihaimaruseac) [inline tensorflow/tools/ci_build/release/requirements_common.txt:5]: Needs to change too

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/tools/ci_build/release/requirements_common.txt (hunk 5 above)
Comment (geetachavan1) [reply tensorflow/tools/ci_build/release/requirements_common.txt:4]: @mihaimaruseac Basically if the nightlies of branch cut day were green, we can be sure that the release build will be green too. correct?

Comment (mihaimaruseac) [reply tensorflow/tools/ci_build/release/requirements_common.txt:4]: If not doing any pinning: green nightly build means green release build only under the assumption that all these packages have not seen a release _between branch cut and final release build_ (or until final patch cut, making this even harder to maintain). This is what caused breakage of RC0 last time even though nightly was green. A package got updated from under us.

If doing the pinning with the version that goes out of range then we have no guarantee that RC0 will be green. We have not run nightly with that package.

This is why I suggested **looking at nightly log for each wheel and seeing what package got installed and pinning to _that specific version_**.

File: tensorflow/tools/ci_build/release/requirements_common.txt [hunk 7]
```diff
@@ -1,22 +1,22 @@
...
-wheel ~= 0.38.1
-wrapt ~= 1.14.1
+numpy == 1.22.4; python_version >= '3.11' # Earliest version for Python 3.11
+opt_einsum == 3.3.0
+protobuf == 3.20.3  # NOTE: Earliest version for Python 3.10
+six == 1.16.0
+termcolor == 2.1.1
+typing_extensions == 3.10.0.2
+wheel == 0.38.4
+wrapt == 1.14.1
```
Comment (mihaimaruseac) [inline tensorflow/tools/ci_build/release/requirements_common.txt:19]: You also have deps from line 33 down

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
Don't forget the other files in the directories too. E.g., https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/release/requirements_mac.txt

File: tensorflow/tools/ci_build/release/requirements_common.txt [hunk 8]
```diff
@@ -1,21 +1,21 @@
...
-absl-py ~= 1.0.0
+absl-py ~= 1.4.0
 astunparse ~= 1.6.3
-flatbuffers ~= 23.1.21
+flatbuffers == 23.3.3
 google_pasta ~= 0.2
 h5py ~= 3.8.0  # Earliest version for Python 3.11
 # TODO(b/262592253): Support older versions of NumPy for Python 3.10 and lower
 # to support TFX. Remove when Apache Beam upgrades to newer NumPy.
 numpy ~= 1.22.0; python_version < '3.11'
```
[the remark this answers is not in the document]
Comment (mihaimaruseac) [reply tensorflow/tools/ci_build/release/requirements_common.txt:11]: Still need to change this numpy.

File: tensorflow/tools/ci_build/release/requirements_common.txt [hunk 9]
```diff
@@ -30,9 +30,9 @@ tb-nightly ~= 2.13.0.a
 tf-estimator-nightly ~= 2.13.0.dev
 
 # Test dependencies
-grpcio ~= 1.49.1 # Earliest version for Python 3.11
-portpicker ~= 1.4.0
-scipy ~= 1.7.2; python_version < '3.11'
+grpcio == 1.49.1 # Earliest version for Python 3.11
+portpicker == 1.4.0
+scipy == 1.7.3; python_version < '3.11'
 scipy ~= 1.9.2; python_version >= '3.11' # Earliest version for Python 3.11
```
Comment (mihaimaruseac) [inline tensorflow/tools/ci_build/release/requirements_common.txt:36]: Don't forget about this scipy

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
Don't forget about the other files.

File: tensorflow/tools/ci_build/release/requirements_common.txt [hunk 10]
```diff
@@ -30,9 +30,13 @@ tb-nightly ~= 2.13.0.a
...
-grpcio ~= 1.49.1 # Earliest version for Python 3.11
-portpicker ~= 1.4.0
-scipy ~= 1.7.2; python_version < '3.11'
+grpcio == 1.49.1 # Earliest version for Python 3.11
+portpicker == 1.4.0
+<<<<<<< HEAD
+scipy ~= 1.7.3; python_version < '3.11'
+=======
+scipy == 1.7.3; python_version < '3.11'
+>>>>>>> refs/remotes/origin/venkat-patch-12
```
Comment (mihaimaruseac) [inline tensorflow/tools/ci_build/release/requirements_common.txt:39]: This has conflict markers

File: tensorflow/tools/ci_build/release/requirements_mac.txt [hunk 11]
```diff
@@ -4,9 +4,9 @@
 certifi ~= 2022.12.07
```
Comment (mihaimaruseac) [inline tensorflow/tools/ci_build/release/requirements_mac.txt:4]: This also needs to change

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt (hunk 2 above)
Comment (mihaimaruseac) [reply tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:14]: Still need to handle this numpy

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt [hunk 12]
```diff
@@ -33,19 +33,19 @@ keras-nightly ~= 2.13.0.dev
 tb-nightly ~= 2.13.0.a
 tf-estimator-nightly ~= 2.13.0.dev
 # Test dependencies
-grpcio ~= 1.49.1 # Earliest version for Python 3.11
-portpicker ~= 1.4.0
-scipy ~= 1.7.2; python_version < '3.11'
+grpcio == 1.49.1 # Earliest version for Python 3.11
+portpicker == 1.4.0
+scipy == 1.7.3; python_version < '3.11'
 scipy ~= 1.9.2; python_version >= '3.11' # Earliest version for Python 3.11
```
Comment (mihaimaruseac) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:39]: Don't forget about this scipy

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt [hunk 13]
```diff
@@ -33,19 +33,19 @@ keras-nightly ~= 2.13.0.dev
...
 # Required for TFLite import from JAX tests
-jax ~= 0.3.24
-jaxlib ~= 0.3.24 # Earliest version for Python 3.11
+jax == 0.3.25
+jaxlib == 0.3.25 # Earliest version for Python 3.11
 # Needs to be addressed. Unblocked 2.4 branchcut cl/338377048
 PyYAML ~= 6.0
```
Comment (mihaimaruseac) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:44]: Needs pin

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt [hunk 14]
```diff
@@ -33,19 +33,19 @@ keras-nightly ~= 2.13.0.dev
...
 # For uploading
 auditwheel ~= 5.0.0
```
Comment (mihaimaruseac) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:46]: Needs pin

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt [hunk 15]
```diff
@@ -33,19 +33,19 @@ keras-nightly ~= 2.13.0.dev
...
-twine ~= 3.6.0
+twine == 3.6.0
 # For user tool scripts
-junitparser ~= 2.2.0
+junitparser == 2.2.0
 lxml ~= 4.9.1
 pylint ~= 2.13.9
```
Comment (mihaimaruseac) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:51]: Need pin

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/tools/ci_build/release/requirements_common.txt [hunk 16]
```diff
@@ -1,22 +1,22 @@
...
-numpy ~= 1.22.0; python_version < '3.11'
-numpy ~= 1.23.2; python_version >= '3.11' # Earliest version for Python 3.11
-opt_einsum ~= 3.3.0
-protobuf ~= 3.20.3  # NOTE: Earliest version for Python 3.10
-six ~= 1.16.0
-termcolor ~= 2.1.1
-typing_extensions ~= 3.10.0.0
-wheel ~= 0.38.1
-wrapt ~= 1.14.1
+numpy ~= 1.22.4; python_version < '3.11'
```
Comment (mihaimaruseac) [inline tensorflow/tools/ci_build/release/requirements_common.txt:11]: ==

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt [hunk 17]
```diff
@@ -4,23 +4,23 @@
...
 # Prefer pinning to the same version as in setup.py for now.
 # This will change in the future.
-absl-py ~= 1.0.0
-astunparse ~= 1.6.3
-flatbuffers ~= 2.0
-google_pasta ~= 0.2
-h5py ~= 3.8.0 # Earliest version for Python 3.11
+absl-py == 1.0.0
+astunparse == 1.6.3
+flatbuffers == 23.1.21
```
Comment (mihaimaruseac) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:9]: This seems wrong. From 2.0 to 23.1.21?

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt (hunk 17 above)
Comment (venkat2469) [reply tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:9]: so can i use the same 2.0 ?

Comment (venkat2469) [reply tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:9]: This is the history for flatbuffers. https://github.com/tensorflow/tensorflow/commits/master/tensorflow/tools/ci_build/release/requirements_common.txt

Comment (mihaimaruseac) [reply tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:9]: You have to use the python wheel version number

Comment (mihaimaruseac) [reply tensorflow/tools/tf_sig_build_dockerfiles/devel.requirements.txt:9]: As for what number to use, you have to see what gets installed in the build log for the last nightly and use that one

**Pull request closed without merging**